### DISCUSSION TITLE
server actions security

### DISCUSSION THREAD

**Discussion by yes-oo** [vercel/next.js#71513] [Help]
### Summary

I'm using server actions in a Next.js app (i.e., functions marked with "use server"). I understand that these functions are executed on the server, but I want to clarify the following:

-     Are the underlying server actions transformed into public API endpoints that can be accessed externally?

-     If so, does this make them publicly accessible by anyone who knows the endpoint?

-     If they are publicly exposed, how can I protect my app from potential abuse, such as attackers making multiple requests to slow down or disrupt the server?

-     Do I need to configure CORS or implement other security mechanisms like rate limiting, authentication, or authorization for server actions?

**Thank you** for the clarification!

**Comment by richardasymmetric** [comment]
> Are the underlying server actions transformed into public API endpoints that can be accessed externally?

Yes

> If so, does this make them publicly accessible by anyone who knows the endpoint?

Yes

> If they are publicly exposed, how can I protect my app from potential abuse, such as attackers making multiple requests to slow down or disrupt the server?
> Do I need to configure CORS or implement other security mechanisms like rate limiting, authentication, or authorization for server actions?

You can. Next has documentation on implementing CORS. For authentication/authorization, you can use something like next-auth. And most this can be done using next-auth. I think next has examples for next-auth and CORS in their examples folder

**Reply by yes-oo** [reply]
thank you for replaying <3 
can you share the examples folder link ?

**Reply by richardasymmetric** [reply]
Here's the next-auth one: https://github.com/vercel/next.js/tree/canary/examples/auth

**Reply by yes-oo** [reply]
Thank you! I’ll wait for someone else to provide a more in-depth answer in the meantime. I really appreciate your time. ❤️
